File: backend/app/services/service_desk_realism.py
"""Replay immutable, bounded workstation operations from the trusted ledger.

The TypeScript renderer interprets the same fixture. Neither snapshots nor
browser-provided evidence fields participate in this replay.
"""

from __future__ import annotations

from copy import deepcopy
import json
from pathlib import Path
import re

CATALOG_VERSION = "realism-v1"
CATALOG_PATH = Path(__file__).resolve().parents[1] / "data/service-desk-realism-v1.json"


def fixture_catalog():
    return json.loads(CATALOG_PATH.read_text())


def read_path(state, path):
    value = state
    for key in path.split("/"):
        value = value.get(key) if isinstance(value, dict) else None
    return value


def write_path(state, path, value):
    keys = path.split("/")
    for key in keys[:-1]:
        state = state.setdefault(key, {})
    state[keys[-1]] = deepcopy(value)


def run_command(state, fixture, command):
    state = deepcopy(state)
    state["realism"]["lastEvidence"] = None
    branches = next(
        (
            rules
            for name, rules in fixture["commands"].items()
            if name.lower() == command.strip().lower()
        ),
        [],
    )
    rule = next(
        (
            candidate
            for candidate in branches
            if all(
                read_path(state, path) == expected
                for path, expected in candidate["when"].items()
            )
        ),
        None,
    )
    if not rule or rule["effect"] == "rejected":
        return state, [], False
    changed_before = state["realism"]["changed"]
    for path, value in rule["set"].items():
        write_path(state, path, value)
    storage = state.get("storage", {})
    if "usageNodeIds" in storage:
        used = storage["fixedUsedBytes"] + sum(
            state["filesystem"]["nodes"][node]["sizeBytes"]
            for node in storage["usageNodeIds"]
        )
        storage["freeBytes"] = storage["capacityBytes"] - used
    if rule["effect"] == "repair":
        state["realism"].update(changed=True, repaired=True)
    if rule["effect"] == "harmful":
        state["realism"]["harmful"] = True
    evidence = rule.get("evidence")
    pre_change = (
        fixture["categories"]["investigation"] + fixture["categories"]["diagnosis"]
    )
    if evidence and not (changed_before and evidence in pre_change):
        state["realism"]["observed"][evidence] = True
        state["realism"]["lastEvidence"] = evidence

    def render(match):
        value = read_path(state, match[1])
        return (
            ", ".join(value)
            if isinstance(value, list)
            else str(value).lower()
            if isinstance(value, bool)
            else str(value)
        )

    return (
        state,
        [re.sub(r"\{([^}]+)\}", render, line) for line in rule["output"]],
        True,
    )


def action_command(fixture, event_type, payload):
    if payload.get("assetTag") != fixture["assetTag"]:
        return None
    if event_type == "remote_desktop.run_terminal_command":
        return (
            payload.get("command") if isinstance(payload.get("command"), str) else None
        )
    if (
        event_type == "remote_desktop.restart_service"
        and payload.get("serviceName") == "Print Spooler"
    ):
        return "Restart-Service -Name Spooler"
    if (
        event_type in {"remote_desktop.start_service", "remote_desktop.stop_service"}
        and payload.get("serviceName") == "Print Spooler"
    ):
        return (
            "Start-Service -Name Spooler"
            if event_type.endswith(".start_service")
            else "Stop-Service -Name Spooler"
        )
    if (
        event_type == "remote_desktop.open_app"
        and payload.get("appId") == "services"
        and fixture["assetTag"] == "NX-2504"
    ):
        return 'sc query "Print Spooler"'
    if event_type == "remote_desktop.explorer_navigate":
        path = payload.get("path", "")
        if not isinstance(path, str):
            return None
        return f"Test-Path {path}" if path.startswith("\\\\") else f"dir {path}"
    return None


def replay(fixture, events):
    state = deepcopy(fixture["initial"])
    for event in events:
        if event.trusted is not True or event.success is not True:
            continue
        command = action_command(fixture, event.event_type, event.payload_json or {})
        if command is not None:
            state, _, _ = run_command(state, fixture, command)
    return state


def transition(fixture, events, event_type, payload):
    command = action_command(fixture, event_type, payload)
    if command is None:
        return None
    state, output, success = run_command(replay(fixture, events), fixture, command)
    return {
        "state": state,
        "output": output,
        "success": success,
        "evidence": state["realism"]["lastEvidence"],
    }

File: service-desk-app/packages/shared/src/service-desk-realism.ts
import catalog from './service-desk-realism-v1.json';
const clone = <T>(value: T): T => JSON.parse(JSON.stringify(value)) as T;

export interface RealismCommand {
  output: string[];
  evidence?: string;
  when: Record<string, unknown>;
  set: Record<string, unknown>;
  effect: string;
}
export interface RealismFixture {
  assetTag: string;
  id: string;
  title: string;
  issue: string;
  hints: string[];
  noteFacts: string[][];
  finalConditions: Record<string, unknown>;
  initial: Record<string, unknown>;
  categories: Record<string, string[]>;
  commands: Record<string, RealismCommand[]>;
  completion: { rootCause: string; whatFixed: string; whyItWorked: string };
}
export const REALISM_FIXTURES: Readonly<Record<string, RealismFixture>> =
  catalog;
export function realismFixture(assetTag: string) {
  return Object.values(REALISM_FIXTURES).find(
    (entry) => entry.assetTag === assetTag,
  );
}
export function realismNoteComplete(fixture: RealismFixture, text: string) {
  return fixture.noteFacts.every((alternatives) =>
    alternatives.some((fact) => text.toLowerCase().includes(fact)),
  );
}

export function readStatePath(state: unknown, path: string): unknown {
  return path
    .split('/')
    .reduce<unknown>(
      (value, key) =>
        value && typeof value === 'object'
          ? (value as Record<string, unknown>)[key]
          : undefined,
      state,
    );
}
export function writeStatePath(state: object, path: string, value: unknown) {
  const keys = path.split('/');
  let target = state as Record<string, unknown>;
  for (const key of keys.slice(0, -1)) {
    target[key] ??= {};
    target = target[key] as Record<string, unknown>;
  }
  target[keys.at(-1)!] = clone(value);
}
export function mergeFixtureState<T extends object>(
  state: T,
  patch: Record<string, unknown>,
): T {
  const next = clone(state);
  const merge = (
    target: Record<string, unknown>,
    source: Record<string, unknown>,
  ) => {
    for (const [key, value] of Object.entries(source)) {
      if (value && typeof value === 'object' && !Array.isArray(value)) {
        target[key] ??= {};
        merge(
          target[key] as Record<string, unknown>,
          value as Record<string, unknown>,
        );
      } else target[key] = clone(value);
    }
  };
  merge(next as Record<string, unknown>, patch);
  return next;
}

/** Bounded declarative operations; never executes a host shell. The server
 * replays these same immutable fixture rules from its trusted ledger. */
export function runRealismCommand<T extends object>(
  state: T,
  fixture: RealismFixture,
  command: string,
) {
  const key = Object.keys(fixture.commands).find(
    (entry) => entry.toLowerCase() === command.trim().toLowerCase(),
  );
  const rule =
    key &&
    fixture.commands[key]?.find((candidate) =>
      Object.entries(candidate.when).every(
        ([path, expected]) =>
          JSON.stringify(readStatePath(state, path)) ===
          JSON.stringify(expected),
      ),
    );
  const next = clone(state);
  writeStatePath(next, 'realism/lastEvidence', null);
  if (!rule)
    return {
      state: next,
      output: [
        'Command or target unavailable in this assigned support scope. Run help for supported syntax.',
      ],
      success: false,
    };
  if (rule.effect === 'rejected')
    return { state: next, output: rule.output, success: false };
  for (const [path, value] of Object.entries(rule.set))
    writeStatePath(next, path, value);
  const usageNodes = readStatePath(next, 'storage/usageNodeIds');
  if (Array.isArray(usageNodes)) {
    const used = usageNodes.reduce<number>(
      (total, id: string) =>
        total + Number(readStatePath(next, `filesystem/nodes/${id}/sizeBytes`)),
      Number(readStatePath(next, 'storage/fixedUsedBytes')),
    );
    writeStatePath(
      next,
      'storage/freeBytes',
      Number(readStatePath(next, 'storage/capacityBytes')) - used,
    );
  }
  if (rule.effect === 'repair') {
    writeStatePath(next, 'realism/changed', true);
    writeStatePath(next, 'realism/repaired', true);
  }
  if (rule.effect === 'harmful') writeStatePath(next, 'realism/harmful', true);
  const preChangeEvidence = [
    ...(fixture.categories.investigation ?? []),
    ...(fixture.categories.diagnosis ?? []),
  ];
  if (
    rule.evidence &&
    !(
      readStatePath(state, 'realism/changed') &&
      preChangeEvidence.includes(rule.evidence)
    )
  ) {
    writeStatePath(next, `realism/observed/${rule.evidence}`, true);
    writeStatePath(next, 'realism/lastEvidence', rule.evidence);
  }
  const output = rule.output.map((line) =>
    line.replace(/\{([^}]+)\}/g, (_, path: string) => {
      const value = readStatePath(next, path);
      return Array.isArray(value) ? value.join(', ') : String(value);
    }),
  );
  return { state: next, output, success: true };
}
